internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go

```diff
@@ -21,96 +21,183 @@ import (
+// every Config.RefreshInterval. To maintain only secrets that are actually needed by the agent, each secret reference
+// expires based on the Config.TTLDelete. During expiration of secret references or actual changes of secret values,
+// the kubernetes_secrets provider calls the ContextProviderComm.Signal() to notify the agent. The cache mechanism
+// can be disabled by setting Config.DisableCache to true.
 func ContextProviderBuilder(logger *logger.Logger, c *config.Config, _ bool) (corecomp.ContextProvider, error) {
-	var cfg Config
+	cfg := defaultConfig()
+
 	if c == nil {
 		c = config.New()
 	}
-	err := c.UnpackTo(&cfg)
+
+	err := c.UnpackTo(cfg)
 	if err != nil {
 		return nil, errors.New(err, "failed to unpack configuration")
 	}
-	return &contextProviderK8sSecrets{
-		logger:       logger,
-		config:       &cfg,
-		secretsCache: make(map[string]*secretsData),
+	return &contextProviderK8SSecrets{
+		logger:  logger,
+		config:  cfg,
+		client:  nil,
+		running: make(chan struct{}),
+		store:   newExpirationCache(cfg.TTLDelete),
 	}, nil
 }
 
-func (p *contextProviderK8sSecrets) Fetch(key string) (string, bool) {
-	if p.config.DisableCache {
-		valid := p.validateKey(key)
-		if valid {
-			return p.fetchSecretWithTimeout(key)
-		} else {
-			return "", false
-		}
-	} else {
-		return p.getFromCache(key)
-	}
-}
-
 // Run initializes the k8s secrets context provider.
-func (p *contextProviderK8sSecrets) Run(ctx context.Context, comm corecomp.ContextProviderComm) error {
+func (p *contextProviderK8SSecrets) Run(ctx context.Context, comm corecomp.ContextProviderComm) error {
 	client, err := getK8sClientFunc(p.config.KubeConfig, p.config.KubeClientOptions)
 	if err != nil {
-		p.logger.Debugf("kubernetes_secrets provider skipped, unable to connect: %s", err)
-		return nil
+		// signal that the provider has initialized
+		close(p.running)
+		p.logger.Debug("kubernetes_secrets provider skipped, unable to connect: ", err.Error())
+		return err
 	}
-	p.clientMx.Lock()
+	p.clientMtx.Lock()
 	p.client = client
-	p.clientMx.Unlock()
+	p.clientMtx.Unlock()
 
 	if !p.config.DisableCache {
-		go p.updateSecrets(ctx, comm)
+		go p.refreshCache(ctx, comm)
 	}
 
+	// signal that the provider has initialized
+	close(p.running)
 	<-comm.Done()
 
-	p.clientMx.Lock()
+	p.clientMtx.Lock()
 	p.client = nil
-	p.clientMx.Unlock()
+	p.clientMtx.Unlock()
+
 	return comm.Err()
 }
 
-func getK8sClient(kubeconfig string, opt kubernetes.KubeClientOptions) (k8sclient.Interface, error) {
-	return kubernetes.GetKubernetesClient(kubeconfig, opt)
+// Fetch returns the secret value for the given key
+func (p *contextProviderK8SSecrets) Fetch(key string) (string, bool) {
+	// Make sure the key has the expected format "kubernetes_secrets.somenamespace.somesecret.value"
+	tokens := strings.Split(key, ".")
+	if len(tokens) > 0 && tokens[0] != "kubernetes_secrets" {
```

Comment:
We should make "kubernetes_secrets" a constant, and use that everywhere we use the string literal.

Reply:
ok fair point const introduced! feel free to unresolve if you spot anything else 🙂